### PR TITLE
FIX ShadowPEFT last-token pooling with left padding

### LINKED ISSUE #3620: [Bug] ShadowPEFT sequence classification pools the wrong token with left padding
### System Info

- PEFT `main`: `9c16ee66cd4c58bd9cdf2d8b4e06c1cf8e8f8efe`
- Affected code: the ShadowPEFT sequence-classification pooling path introduced in #3354.

### Who can help?

@BenjaminBossan

### Reproduction

ShadowPEFT documents `_pool_last_token` as selecting the last non-padding representation, but it derives the index as `attention_mask.sum(dim=1) - 1`:

```python
import torch

from peft.tuners.shadow.model import _pool_last_token

hidden = torch.tensor(
    [
        [[10.0], [20.0], [30.0], [40.0]],
        [[10.0], [20.0], [30.0], [40.0]],
    ]
)
attention_mask = torch.tensor(
    [
        [0, 0, 1, 1],
        [0, 1, 1, 1],
    ]
)

print(_pool_last_token(hidden, attention_mask).squeeze(-1))
```

The current index expression selects positions 1 and 2, producing `[20, 30]`. Both are incorrect for these left-padded rows. The last non-padding position is 3 in each row, so the expected result is `[40, 40]`.

The same count-based implementation appears twice:

- the helper used by `ShadowModel.shadow_auxiliary_loss`, so the auxiliary sequence-classification loss trains against a padding position;
- `DetachedShadowModel._pool_last_token`, so `unload_shadow()` inference returns logits from the wrong position.

Transformers sequence-classification models intentionally select the rightmost non-padding position to support both left and right padding:

https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_layers.py

Tokenizer configuration also exposes `padding_side` as a supported choice:

https://huggingface.co/docs/transformers/main_classes/tokenizer

I searched PEFT issues, pull requests, and history for ShadowPEFT padding, classification pooling, left padding, and `_pool_last_token`. I found no overlapping report or implementation.

### Expected behavior

ShadowPEFT sequence classification should select the rightmost position whose attention-mask value is nonzero, matching Transformers for both left- and right-padded batches.

The proposed scope is one shared pooling helper used by both training and detached inference, with focused regressions for left and right padding. Causal-language-model and cache behavior would remain unchanged.

Would you like me to prepare that focused PR? I will wait for explicit approval before changing code.

AI assistance was used during the source audit and duplicate search. I reviewed the complete affected path and independently checked the index calculation against the documented Transformers behavior.

### PR DESCRIPTION
**Pull request by Cunninger**
Fixes #3620

`_pool_last_token` indexed with `attention_mask.sum(dim=1) - 1`, which is the last **non-padding** token only for **right** padding. With left padding it lands in the pad region.

The helper now weights the mask by position and takes `argmax`, which selects the last non-pad token for left-padded, right-padded, and mixed batches. The same logic is used in both `peft.tuners.shadow.model._pool_last_token` and `DetachedShadowModel._pool_last_token`.

## Tests

```
pytest tests/test_shadow.py::TestShadowSequenceClassification::test_pool_last_token_left_padding tests/test_shadow.py::TestShadowSequenceClassification::test_pool_last_token_right_padding tests/test_shadow.py::TestShadowSequenceClassification::test_pool_last_token_without_mask_uses_final_position -v
```

All three passed locally.

## AI assistance

AI assistance was used to draft this change. I reviewed every line, reproduced the left-padding pooling bug from #3620, and ran the tests above. Coordination comment: https://github.com/huggingface/peft/issues/3620#issuecomment-5467966838

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
The poster of the issue proposed to fix the problem. Please don't just open PRs without approval.
